modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/AbstractExternalResult.java

```diff
@@ -0,0 +1,374 @@
+        }
+    }
+
+    /**
+     * Used for {@link ResultInterface#createShallowCopy(org.h2.engine.SessionInterface)} only.
+     * @param parent Parent result.
+     */
+    protected AbstractExternalResult(AbstractExternalResult parent) {
+        log = parent.log;
+        size = parent.size;
+        file = parent.file;
+        fileIo = parent.fileIo;
+        this.parent = parent;
+        memTracker = parent.memTracker;
+
+        if (memTracker != null)
+            memTracker.registerCloseListener(this::close);
+    }
+
+    /**
+     * Reads full row from file starting from the given position.
+     * @param addr Given position.
+     *
+     * @return Row.
+     */
+    protected Value[] readRowFromFile(long addr) {
+        setFilePosition(addr);
+
+        return readRowFromFile();
+    }
+
+    /**
+     * Reads full row from file starting from the current position.
+     *
+     * @return Row.
+     */
+    protected Value[] readRowFromFile() {
+        ByteBuffer hdr = readRowHeaderFromFile();
+
+        int rowLen = hdr.getInt();
+        int colCnt = hdr.getInt();
+
+        if (colCnt == -1) {
+            setFilePosition(currentFilePosition() + rowLen);
+
+            return null; // Row has been deleted.
+        }
+
+        ByteBuffer rowBytes = readDataFromFile(rowLen);
+
+        Data buff = Data.create(null, rowBytes.array(), true);
+
+        Value[] row = new Value[colCnt];
+
+        for (int i = 0; i < colCnt; i++)
+            row[i] = buff.readValue();
+
+        return row;
+    }
+
+    /**
+     * Reads row header starting from the current position.
+     *
+     * @return Buffer with row header
+     */
+    @NotNull protected ByteBuffer readRowHeaderFromFile() {
+        return readDataFromFile(ROW_HEADER_SIZE);
+    }
+
+    /**
+     * Reads data from file.
+     *
+     * @param rowLen Data length to read.
+     * @return Byte buffer with data.
+     */
+    @NotNull protected ByteBuffer readDataFromFile(int rowLen) {
+        try {
+            ByteBuffer rowBytes = ByteBuffer.allocate(rowLen); // 2 x integer for length in bytes and columns count.
+
+            while (rowBytes.hasRemaining()) {
+                int bytesRead = fileIo.read(rowBytes);
+
+                if (bytesRead <= 0)
+                    throw new IOException("Can not read data from file: " + file.getAbsolutePath() +
+                        ", curPos=" + fileIo.position() + ", rowLen=" + rowLen + "]");
+            }
+
+            rowBytes.flip();
+
+            return rowBytes;
+        }
+        catch (IOException e) {
+            close();
+
+            throw new IgniteException("Failed to read query result the from spill file.", e);
+        }
+    }
+
+    /**
+     * @return New data buffer.
+     */
+    @NotNull protected Data createDataBuffer() {
+        return Data.create(null, INIT_BUF_SIZE, true); // TODO init size
```

Comment:
Let's remove "TODO" or specify a ticket number.
BTW, H2 uses Constants.DEFAULT_PAGE_SIZE = 4096 as default buffer size for rows.

Reply:
Done